File: haskell/2015/day13b/src/Main.hs
module Main (main) where

import Data.Int (Int32)
import Data.List
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Text.Parsec
import Text.Parsec.String

type HappinessTable = Map.Map (String, String) Int

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

file :: Parser [(String, String, Int)]
file = line `endBy` newline

line :: Parser (String, String, Int)
line = do
       person1 <- many1 letter
       _ <- string " would "
       action <- string "gain" <|> string "lose"
       spaces
       amountTxt <- many1 digit
       _ <- string " happiness units by sitting next to "
       person2 <- many1 letter
       _ <- string "."
       let amount = case action of
                        "gain" -> read amountTxt
                        "lose" -> -(read amountTxt)
                        _ -> 0
       return (person1, person2, amount)

buildHapTable :: [(String, String, Int)] -> HappinessTable
buildHapTable = foldl' (\acc (p1, p2, h) -> Map.insert (p1, p2) h acc) Map.empty

injectSelf :: HappinessTable -> [String] -> HappinessTable
injectSelf hapTable people =
    foldl' (\acc person -> Map.insert ("Self", person) 0 (Map.insert (person, "Self") 0 acc)) hapTable people

computeScore :: HappinessTable -> [String] -> Int
computeScore hapTable people =
    sum [ hapTable Map.! (p, l) + hapTable Map.! (p, r) | (p, l, r) <- triples people ]
  where
    triples xs = zip3 xs (last xs : init xs) (tail xs ++ [head xs])

process :: String -> Int32
process content =
    case parse file "" content of
        Left err -> error (show err)
        Right arrangements ->
            let people = Set.toList $ Set.fromList [ p1 | (p1, _, _) <- arrangements ]
                hapTable = injectSelf (buildHapTable arrangements) people
                (fixed:rest) = people ++ ["Self"]
            in fromIntegral $ maximum $ map (computeScore hapTable . (fixed:)) (permutations rest)

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            content <- readFile filename
            let result = process content
            putStrLn $ "result = " ++ show result
        _ -> usage progname

File: rust/2015/day13b/src/main.rs
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;

use itertools::Itertools;
use regex::Regex;

type HappinessTable = HashMap<(String, String), i32>;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn parse_file(filename: &str) -> io::Result<HappinessTable> {
    let mut table: HappinessTable = HashMap::new();
    let re = Regex::new(r"(\w+) would (gain|lose) (\d+) happiness units by sitting next to (\w+).")
        .unwrap();
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    for line_result in reader.lines() {
        let line = line_result?;
        if let Some(caps) = re.captures(&line) {
            let p1 = String::from(caps.get(1).unwrap().as_str());
            let p2 = String::from(caps.get(4).unwrap().as_str());
            let amt = match caps.get(2).unwrap().as_str() {
                "gain" => caps.get(3).unwrap().as_str().parse::<i32>().unwrap(),
                "lose" => -(caps.get(3).unwrap().as_str().parse::<i32>().unwrap()),
                _ => unreachable!(),
            };
            table.insert((p1, p2), amt);
        }
    }
    Ok(table)
}

fn compute_score(table: &HappinessTable, people: &[String]) -> i32 {
    let n = people.len();
    let mut total = 0;
    for i in 0..n {
        let p = &people[i];
        let l = &people[(i + n - 1) % n];
        let r = &people[(i + 1) % n];
        total += table.get(&(p.clone(), l.clone())).unwrap_or(&0);
        total += table.get(&(p.clone(), r.clone())).unwrap_or(&0);
    }
    total
}

fn inject_self(table: &mut HappinessTable) {
    let mut people_set: HashSet<String> = HashSet::new();
    table.keys().for_each(|(p1, p2)| {
        people_set.insert(p1.clone());
        people_set.insert(p2.clone());
    });
    let people: Vec<String> = people_set.into_iter().collect();
    for person in people.iter() {
        table.insert((String::from(person), String::from("Self")), 0);
        table.insert((String::from("Self"), String::from(person)), 0);
    }
}

fn process_file(filename: &str) -> io::Result<i32> {
    let mut table = parse_file(filename)?;
    inject_self(&mut table);
    let mut people_set: HashSet<String> = HashSet::new();
    table.keys().for_each(|(p1, p2)| {
        people_set.insert(p1.clone());
        people_set.insert(p2.clone());
    });
    let mut people: Vec<String> = people_set.into_iter().collect();
    people.sort();
    let (fixed, rest) = people.split_first().unwrap();
    let result = rest
        .iter()
        .permutations(rest.len())
        .map(|perm| {
            let mut arrangement = vec![fixed.clone()];
            arrangement.extend(perm.into_iter().cloned());
            compute_score(&table, &arrangement)
        })
        .max()
        .unwrap_or(0);
    Ok(result)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}
